toolchain/check/handle_function.cpp

```diff
@@ -44,9 +44,13 @@ auto HandleParseNode(Context& context, Parse::ReturnTypeId node_id) -> bool {
   // Propagate the type expression.
   auto [type_node_id, type_inst_id] = context.node_stack().PopExprWithNodeId();
   auto type_id = ExprAsType(context, type_node_id, type_inst_id).type_id;
-  // TODO: Use a dedicated instruction rather than VarStorage here.
-  context.AddInstAndPush<SemIR::VarStorage>(
-      node_id, {.type_id = type_id, .name_id = SemIR::NameId::ReturnSlot});
+  auto return_slot_id = context.AddPatternInst<SemIR::ReturnSlotPattern>(
+      node_id, {.type_id = type_id});
+  auto param_pattern_id = context.AddPatternInst<SemIR::ParamPattern>(
+      node_id, {.type_id = type_id,
+                .subpattern_id = return_slot_id,
+                .runtime_index = SemIR::RuntimeParamIndex::Unknown});
```

Comment:
Looking at the generated SemIR, it looks like the `Param` inst we create for this pattern is treated as a value expression. That doesn't seem correct to me. Is there some way we can fix that? (Are we going to need a second instruction kind for an `OutParam` or something like that?)

Reply:
[We also discussed this in the toolchain meeting today; I wrote up the comment below prior to that meeting.]

For option (2), I think I'd want something specific to `return` in the instruction name, given that we don't have the return slot instruction any more. But I think I like separating the return parameter from the return slot, at least in theory, and that might open up a way to represent things like a return value that's partly returned in memory and partly returned through registers, or for which we pass multiple parameters for locations of parts of the return value. So I think that leads me to prefer option (1).

Additionally I think we'll want something similar to distinguish `var` parameters from `let` parameters, because the same expr category decision is going to be needed there too. So I think we will end up with an `AnyParam` category anyway. In principle we could use the same instruction kind for both `var` parameters and return slot parameters, but in practice I think we should distinguish them. (Also, they're not the same, I suppose: for a type that's returned in registers, a `RefParam` should be lowered as a pointer parameter but an `OutParam` should not add a parameter to the LLVM function at all.)

Reply:
OK, I've added a TODO to implement option 1.